discojs/src/models/tfjs.ts

```diff
@@ -30,69 +39,58 @@ export class TFJS extends Model {
   }
 
   override async *train(
-    trainingData: tf.data.Dataset<tf.TensorContainer>,
-    validationData?: tf.data.Dataset<tf.TensorContainer>,
+    trainingDataset: Dataset<Batched<DataTypeToPreprocessedLabeled[D]>>,
+    validationDataset?: Dataset<Batched<DataTypeToPreprocessedLabeled[D]>>,
   ): AsyncGenerator<BatchLogs, EpochLogs> {
-    const batches = await trainingData.iterator(); // tf.LazyIterator isn't an AsyncGenerator
     let batchesLogs = List<BatchLogs>();
-    for (let batchNumber = 0; true; batchNumber++) {
-      const iteration = await batches.next();
-      if (iteration.done) break;
-      const batch = iteration.value;
+    for await (const [batch, batchNumber] of trainingDataset.zip(Range())) {
 
       const batchLogs = {
         batch: batchNumber,
         ...(await this.#runBatch(batch)),
       };
-      tf.dispose(batch);
 
       yield batchLogs;
       batchesLogs = batchesLogs.push(batchLogs);
     }
 
-    const validation = validationData && (await this.#evaluate(validationData));
+    const validation = validationDataset && (await this.#evaluate(validationDataset));
     return new EpochLogs(batchesLogs, validation);
   }
 
   async #runBatch(
-    batch: tf.TensorContainer,
+    batch: Batched<DataTypeToPreprocessedLabeled[D]>,
   ): Promise<Omit<BatchLogs, "batch">> {
-    let logs: tf.Logs | undefined;
-    await this.model.fitDataset(tf.data.array([batch]), {
+    const { xs, ys } = this.#batchToTF(batch);
+
+    const { history } = await this.model.fit(xs, ys, {
```

Comment:
because GPT is misusing TFJS quite a lot :face_with_diagonal_mouth: instead of reusing TFJS's training loop, it is redoing it, by overriding `fitDataset` (and forgetting to also implement `fit`)… I don't think there is any real reason.
if we want to put a bunch of hours on fixing it, we can probably get rid of it entirely (it remaind of smth I tried to do a while back) but I'm not sure that the gain will be enormous, especially when ONNX would be better,faster,stronger